src/providers/script/base.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * Notation:
+ *   - A "word" is a 32-bit interger
+ */
+import { ByteBuffer } from 'dojo-core/encoding';
+
+/**
+ * A script hash function
+ */
+export interface ScriptHash {
+	(data: ByteBuffer): ByteBuffer;
+	blockSize: number;
+}
+
+/**
+ * A general math function
+ */
+export interface MathFunction {
+	(...inputs: number[]): number
+}
+
+/**
+ * Add a list of words, with rollover
+ */
+export function addWords(...words: number[]): number {
+	const numWords = words.length;
+	let sum = words[0];
+	for (let i = 1; i < numWords; i++) {
+		const a = sum;
+		const b = words[i];
+		const low = (a & 0xFFFF) + (b & 0xFFFF);
+		const high = (a >> 16) + (b >> 16) + (low >> 16);
+		sum = (high << 16) | (low & 0xFFFF);
+	}
+	return sum;
+}
+
+/**
+ * Specify the endian-ness of a integer values
+ */
+export enum Endian {
+	Little = 0,
+	Big = 1
+}
+
+/**
+ * Convert an array of bytes to an array of 32-bit words. Words are assumed to be encoded in little-endian format (low
+ * bytes are at lower indices).
+ */
+export function bytesToWords(bytes: ByteBuffer, endian: Endian = Endian.Big): number[] {
+	const numWords = Math.ceil(bytes.length / 4);
+	const words = new Array(numWords);
```

Comment:
I could go either way on this.  It's not in our style guide, but from what I vaguely recall, the rule, when it exists, is usually specifically to discourage using `new Array()` when the intent is identical to `[ ... ]`, particularly since someone could mistakenly try `new Array(number)` when they mean `[ number ]` but will end up with an array of `number` undefined values.  in this case though, the intent *is* to preallocate.